Gossip RTS membership between servers. Notify and install membership changes to RTSs. Implement actor placement by consistent hashing

aagapi opened the pull request:
This PR includes the following changes:
  - As RTSs connect to DB servers, the DB servers: 1) maintain connection status of RTSs; 2) Include information about connected RTSs (including metadata about them and liveness status) in their ongoing gossip agreement protocols with the other DB servers
  - Agreed gossiped views between DB servers now include membership of connected RTSs (in addition to DB server membership as before)
  - As new views are learned by DB servers, these are in turn pushed to and installed into all connected RTS clients, which serves 2 purposes: 1) RTSs find out about status changes in server membership to be able to maintain an up-to-date view of the DB server ring; 2) RTSs find out about each other (which other RTSs participate in the system == the RTS ring)
  - The root actor in the RTS notifies the RTS layer about the set of actors in the system (either as they are created for the first time, or as they are found in the DB by the recovery process) == the actor ring
  - Using the RTS ring and actor ring information, RTSs now independently and deterministically assign actors to RTSs; in the current implementation consistent hashing with a collision free hash function is used for this process
  - As new RTS membership views get installed by the gossip layers in RTSs, the actor-to-RTS mapping is automatically maintained

aagapi commented on test/test_db.py:
```diff
@@ -110,7 +110,7 @@ def __init__(self, idx, base_port=32000):
         self.name = f"Db{idx}"
         self.port = base_port + self.idx
-        self.seed_port = base_port + 100
+        self.seed_port = base_port + 2000
```
This is just a temporary workaround until RPC ports (as opposed to gossip ports) are learned dynamically. There is a hardcoded offset of 2000 that I temporarily assumed for the workaround until this is implemented, and I changed this to match the testing script for it to pass, since I assumed the offset in the testing script was arbitrary. Will roll back this change and fix the workaround on my side.

aagapi commented on rts/rts.c:
```diff
@@ -1908,7 +1919,11 @@ int main(int argc, char **argv) {
-        db = get_remote_db(ddb_replication);
+        int random_id;
+		FASTRAND(&seed, random_id);
+		random_id = 5000 + random_id % 100;
```
This is again just temporary for my own testing purposes with a couple of RTS nodes, it will not remain in the final PR. We will not have random ids in the final version as we cannot afford any collisions, however improbable. I will replace this with a config param.

plajjan replied:
I like the random IDs. We can have a config option too but I think random IDs is a good default because then the user won't have to 

To avoid collisions, I think we should have a duplicate ID detection mechanism. It is the only way to be safe; leaving it to the user to specify with a configuration option does not avoid the problem, it only moves responsibility from us to the end user, which IMHO is poor UX. I want to help the user and help them prevent data loss / faulty behavior. Duplicate ID detection would be the best way to avoid that.

I don't think this needs to land in the initial PR though, we can build it on top later if you prefer.

Without knowing anything about the gossip protocol, it seems quite simple; just connect and receive information about the other nodes, check if there already is a node with our ID; if there is, exit with an error? Otherwise continue and announce ourselves?  But you'll probably tell me there is a very early hello message where we announce ourselves with our ID? ;)

aagapi replied:
In principle it is possible to do duplicate detection of the RTS IDs using the gossip layer (or even easier, to generate unique IDs for RTSs automatically in the gossip agreement process - see below), but it would require some re-coding.
Currently IDs get propagated upon connection to the database from clients -> gossip membership agreement between servers (integrated with the existing gossip layer) -> pushed back to all other clients. In other words, a client (RTS) must have joined (and specified its ID) before it learns info about the other ones via gossip.
It is possible to have it join with essentially no ID, then have one assigned for it by the gossip / agreement layer itself automatically, so that the user has to specify nothing. I think this would be preferable to random ID picking by clients + collision detection, since the latter would be a more complicated solution achieving the same thing but requiring us to handle more corner cases (for instance there is no way to guarantee 2 RTSs joining very close in time to each other will have learned about the other one immediately via gossip, in which case we need a mechanism to change an already assigned ID. However unlikely, it is a case we still need to handle in code).
Generating the IDs during the agreement process is will be easier and require less message roundtrips, but I will do this in a separate PR, let's consider this PR v1.
For now I have added configuration params for: rts ID, rts hostname, as well as rack ID and datacenter ID. A unique ID is generated from the combination of hostname + rts ID, allowing the user to either deploy RTSs on different hosts, or several of them on the same host but with different rts IDs. In principle, if people use something like Kubernetes or other automatic framework for deployment rather than doing it by hand, it is likely they have a standard way of getting hostnames and ids dynamically upon each deployment.
But I agree it is better to have less configuration, so we can remove the rts_id parameter once we have a way to generate them automatically as above. We should still keep the rack_id and datacenter_id parameters though to support more advanced placement strategies based on various failures scenarios, such as partitions vs node failures.

aagapi commented on backend/failure_detector/db_queries.c:
```diff
@@ -2124,14 +2226,55 @@ void free_client_msg(ClientMessage * cm)
 	{
-		fprintf(stderr, "error unpacking server message\n");
-	    return 1;
+		fprintf(stderr, "error unpacking client message\n");
```
Will do. Yes, some places still remain with fprintf(stderr), I used the logging lib in general. I will fix in the final PR, good point.

aagapi commented:
Converting this PR to draft, as I am looking into some crashes I noticed on MacOS. They may or may not be related to the content of this PR, I am currently investigating.
Also, there are still a few things I want to add before I make the merge final, some of which @plajjan noticed above (e.g. using a configured ID for RTSs versus randomly generated ones, which was just a temporary stand-in I've been using during development). I will un-draft the PR when the branch is final and MacOS tests pass.

plajjan commented on rts/rts.c:
```diff
@@ -1808,6 +1827,18 @@ int main(int argc, char **argv) {
+                rts_node_id = atoi(optarg);
+                break;
+            case 'R':
+            		rts_rack_id = atoi(optarg);
```
indent (tabs) ;)

plajjan commented on rts/rts.c:
```diff
@@ -1808,6 +1827,18 @@ int main(int argc, char **argv) {
+            		rts_rack_id = atoi(optarg);
+                break;
+            case 'D':
+            		rts_dc_id = atoi(optarg);
```
indent (tabs) ;)

plajjan commented on rts/rts.c:
```diff
@@ -1907,21 +1938,24 @@ int main(int argc, char **argv) {
-            int port = ddb_port;
-            char *colon = strchr(host, ':');
+        for (int i=0; i<ddb_no_host; i++) {
+        		seed_hosts[i] = strdup(ddb_host[i]);
+        		seed_ports[i] = ddb_port;
```
more tab indent... I'll stop commenting for tab indent :stuck_out_tongue_winking_eye:

aagapi replied:
Fixed. Note that there are some tabs in that file that pre-date the commits in this PR, most also coming from DB client code. I'd rather fix that in a separate "indentation fix" PR to keep the set of lines in the RTS functionally affected by this PR clear. I've fixed all tabs that were introduced by changes in this PR.

aagapi commented:
Updates:
 - Fixed segfault occurring on MacOS
 - Used config params for configuration of rts hostname, rts ID, as well as rack ID and datacenter ID (see comment above in reply to @plajjan for more details and possible follow-up).
 - Dynamic learning of RPC ports for DB servers (in addition to gossip ports) via gossip still remains to be done as future work to allow gossip/RPC ports to be independently configured per server, but will be handled in a separate PR. Current workaround assumes a hardcoded difference of 1 between RPC and gossip ports, which needs to be considered in any test/deployment scripts (this was currently done for test/test_db.py by @plajjan)
 - Fixed tab indenting in rts.c changes and migrated db_queries.c to use the logging lib
